pkg/image/controller/imagestream_controller.go

```diff
@@ -25,25 +53,131 @@ type Notifier interface {
-type NotifierFunc func(stream *api.ImageStream)
+type ImageStreamController struct {
+	// image stream client
+	isNamespacer client.ImageStreamsNamespacer
+
+	// queue contains replication controllers that need to be synced.
+	queue workqueue.RateLimitingInterface
+
+	// lister can list/get image streams from a shared informer's cache
+	lister imageStreamLister
+	// listerSynced makes sure the is store is synced before reconciling streams
+	listerSynced cache.InformerSynced
 
-// Importing adapts NotifierFunc to Notifier
-func (fn NotifierFunc) Importing(stream *api.ImageStream) {
-	fn(stream)
+	// notifier informs other controllers that an import is being performed
+	notifier Notifier
 }
 
-// tagImportable is true if the given TagReference is importable by this controller
-func tagImportable(tagRef api.TagReference) bool {
-	if tagRef.From == nil {
-		return false
+func (c *ImageStreamController) SetNotifier(n Notifier) {
+	c.notifier = n
+}
+
+// Run begins watching and syncing.
+func (c *ImageStreamController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	glog.Infof("Starting image stream controller")
+
+	// Wait for the stream store to sync before starting any work in this controller.
+	if !cache.WaitForCacheSync(stopCh, c.listerSynced) {
+		utilruntime.HandleError(fmt.Errorf("timed out waiting for caches to sync"))
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+	glog.Infof("Shutting down image stream controller")
+}
+
+func (c *ImageStreamController) addImageStream(obj interface{}) {
+	if stream, ok := obj.(*api.ImageStream); ok {
+		c.enqueueImageStream(stream)
+	}
+}
+
+func (c *ImageStreamController) updateImageStream(old, cur interface{}) {
+	curStream, ok := cur.(*api.ImageStream)
+	if !ok {
+		return
+	}
+	oldStream, ok := old.(*api.ImageStream)
+	if !ok {
+		return
 	}
-	if tagRef.From.Kind != "DockerImage" || tagRef.Reference {
+	// we only compare resource version, since deeper inspection if a stream
+	// needs to be re-imported happens in syncImageStream
+	//
+	// FIXME: this will only be ever true on cache resync
+	if curStream.ResourceVersion == oldStream.ResourceVersion {
+		return
+	}
+	c.enqueueImageStream(curStream)
+}
+
+func (c *ImageStreamController) enqueueImageStream(stream *api.ImageStream) {
+	key, err := kcontroller.KeyFunc(stream)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("Couldn't get key for image stream %#v: %v", stream, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *ImageStreamController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *ImageStreamController) processNextWorkItem() bool {
+	key, quit := c.queue.Get()
+	if quit {
 		return false
 	}
+	defer c.queue.Done(key)
+
+	stream, err := c.getByKey(key.(string))
+	if err == nil && stream != nil {
+		glog.V(3).Infof("Queued import of stream %s/%s...", stream.Namespace, stream.Name)
```

Comment:
It's slightly different than what we've had before. But this is [this](https://github.com/openshift/origin/pull/13835/files#diff-0da76d25b3dc65e4fc80f1ab22aa745aL192) and it was 3 back then.